File: 101-symmetric-tree/symmetric-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* l, TreeNode* r) {
        if (!l || !r) 
            return !l && !r;
        if (l->val != r->val)
            return false;
        return isSymmetric(l->left, r->right) && isSymmetric(l->right, r->left);
    }
    bool isSymmetric(TreeNode* root) {
        return isSymmetric(root->left, root->right);
    }
};

File: 101-symmetric-tree/symmetric-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSymmetric(root *TreeNode) bool {
    if root == nil {
        return true
    }

    var dfs func(left *TreeNode, right *TreeNode) bool
    dfs = func(left *TreeNode, right *TreeNode) bool {
        if left == nil && right == nil {
            return true
        }

        if left == nil || right == nil {
            return false
        }

        if left.Val != right.Val {
            return false
        }

        return dfs(left.Left, right.Right) && dfs(right.Left, left.Right)
    }

    return dfs(root.Left, root.Right)
}